### PR TITLE
Make sure TimeIntervalRounding is monotonic for increasing dates

### PR DESCRIPTION
**Pull request by cbuescher**
Currently there are cases when using TimeIntervalRounding#round() that if `date1 < date2` we get `round(date2) < round(date1)`. These errors can happen when using a non-fixed time zone and the values to be rounded are slightly after a time zone offset change (e.g. DST transition).

Here is an example for the "CET" time zone with a 45 minute rounding interval. The dates to be rounded are on the left (with utc time stamp), the rounded values on the right. The error case is marked:

| orig. date | rounded |
| --- | --- |
| 2011-10-30T01:40:00.000+02:00 1319931600000 |  2011-10-30T01:30:00.000+02:00 1319931000000 |
| 2011-10-30T02:02:30.000+02:00 1319932950000 |  2011-10-30T01:30:00.000+02:00 1319931000000 |
| 2011-10-30T02:25:00.000+02:00 1319934300000 |  2011-10-30T02:15:00.000+02:00 1319933700000 |
| 2011-10-30T02:47:30.000+02:00 1319935650000 |  2011-10-30T02:15:00.000+02:00 1319933700000 |
| 2011-10-30T02:10:00.000+01:00 1319937000000 |  2011-10-30T01:30:00.000+02:00 1319931000000 < previous rounded value |
| 2011-10-30T02:32:30.000+01:00 1319938350000 |  2011-10-30T02:15:00.000+01:00 1319937300000 |
| 2011-10-30T02:55:00.000+01:00 1319939700000 |  2011-10-30T02:15:00.000+01:00 1319937300000 |
| 2011-10-30T03:17:30.000+01:00 1319941050000 |  2011-10-30T03:00:00.000+01:00 1319940000000 |

We should correct this by detecting that we are crossing a transition when rounding, and in that case pick the largest valid rounded value before the transition.

This change adds this correction logic to the rounding function and adds this invariant to the randomized TimeIntervalRounding tests. Also adding the example test case from above (with corrected behaviour) for illustrative purposes.

### CODE REVIEW

**Comment by cbuescher** [comment]
@jpountz @jimferenczi this is a small spin-off of a larger change I have locally, maybe one of you can take a look.

File: core/src/test/java/org/elasticsearch/common/rounding/TimeZoneRoundingTests.java [hunk 1]
```diff
@@ -328,30 +332,71 @@ public void testIntervalRoundingRandom() {
...
                 }
-            } catch (AssertionError e) {
-                logger.error("Rounding error at {}, timezone {}, interval: {},", new DateTime(date, tz), tz, interval);
-                throw e;
             }
         }
     }
 
     /**
+     * Test that rounded values are always greater or eaual to last rounded value if date is increasing.
```
Comment (jpountz) [inline core/src/test/java/org/elasticsearch/common/rounding/TimeZoneRoundingTests.java]: s/eaual/equal/

Comment (cbuescher) [inline core/src/test/java/org/elasticsearch/common/rounding/TimeZoneRoundingTests.java]: Thanks, corrected that.

**Comment by jpountz** [comment]
LGTM

**Comment by cbuescher** [comment]
@jpountz thanks for the review

**Comment by cbuescher** [comment]
Merged to 2.4 with 31ade93.

**Pull request merged**